File: apps/backend/apple_music.py
"""
Apple Music API helpers and routes.
"""
from __future__ import annotations

import os
import time
from typing import Optional

import httpx
import jwt
from fastapi import APIRouter, Header, HTTPException

APPLE_MUSIC_API_BASE = "https://api.music.apple.com"
APPLE_MUSIC_TOKEN_MAX_TTL_SECONDS = 60 * 60 * 24 * 180  # 6 months

router = APIRouter(prefix="/apple-music", tags=["apple-music"])

_token_cache: dict[str, int | str] = {}


def _load_private_key() -> str:
    key_pem = os.getenv("APPLE_MUSIC_PRIVATE_KEY")
    key_path = os.getenv("APPLE_MUSIC_PRIVATE_KEY_PATH")

    if key_pem:
        return key_pem.replace("\\n", "\n").strip()

    if key_path:
        try:
            with open(key_path, "r", encoding="utf-8") as handle:
                return handle.read().strip()
        except OSError as exc:
            raise RuntimeError(f"Failed to read APPLE_MUSIC_PRIVATE_KEY_PATH: {exc}") from exc

    raise RuntimeError("APPLE_MUSIC_PRIVATE_KEY or APPLE_MUSIC_PRIVATE_KEY_PATH is required")


def _generate_developer_token() -> tuple[str, int]:
    team_id = os.getenv("APPLE_MUSIC_TEAM_ID")
    key_id = os.getenv("APPLE_MUSIC_KEY_ID")
    ttl_seconds = int(os.getenv("APPLE_MUSIC_TOKEN_TTL_SECONDS", "3600"))

    if not team_id or not key_id:
        raise RuntimeError("APPLE_MUSIC_TEAM_ID and APPLE_MUSIC_KEY_ID are required")

    ttl_seconds = min(max(ttl_seconds, 60), APPLE_MUSIC_TOKEN_MAX_TTL_SECONDS)
    now = int(time.time())
    expires_at = now + ttl_seconds

    payload = {
        "iss": team_id,
        "iat": now,
        "exp": expires_at,
    }
    headers = {"alg": "ES256", "kid": key_id}

    private_key = _load_private_key()
    token = jwt.encode(payload, private_key, algorithm="ES256", headers=headers)
    return token, expires_at


def get_developer_token() -> tuple[str, int]:
    cached_token = _token_cache.get("token")
    cached_exp = _token_cache.get("exp")
    now = int(time.time())

    if cached_token and cached_exp and now < int(cached_exp) - 60:
        return str(cached_token), int(cached_exp)

    token, exp = _generate_developer_token()
    _token_cache["token"] = token
    _token_cache["exp"] = exp
    return token, exp


def _build_headers(user_token: Optional[str] = None) -> dict[str, str]:
    token, _ = get_developer_token()
    headers = {"Authorization": f"Bearer {token}"}
    if user_token:
        headers["Music-User-Token"] = user_token
    return headers


async def _apple_music_get(
    path: str,
    params: Optional[dict[str, str | int]] = None,
    user_token: Optional[str] = None,
) -> dict:
    url = f"{APPLE_MUSIC_API_BASE}/{path.lstrip('/')}"
    headers = _build_headers(user_token=user_token)
    async with httpx.AsyncClient(timeout=15) as client:
        response = await client.get(url, headers=headers, params=params)
    if response.status_code >= 400:
        raise HTTPException(status_code=response.status_code, detail=response.text)
    return response.json()


@router.get("/developer-token")
async def developer_token():
    """Return a signed Apple Music developer token."""
    try:
        token, exp = get_developer_token()
    except RuntimeError as exc:
        raise HTTPException(status_code=500, detail=str(exc)) from exc
    return {"token": token, "expires_at": exp}


@router.get("/catalog/search")
async def catalog_search(
    term: str,
    types: str = "songs",
    storefront: str = "us",
    limit: int = 10,
    offset: int = 0,
):
    """Search the Apple Music catalog."""
    params = {
        "term": term,
        "types": types,
        "limit": limit,
        "offset": offset,
    }
    return await _apple_music_get(f"v1/catalog/{storefront}/search", params=params)


@router.get("/catalog/songs/{song_id}")
async def catalog_song(song_id: str, storefront: str = "us"):
    """Fetch a song by catalog ID."""
    return await _apple_music_get(f"v1/catalog/{storefront}/songs/{song_id}")


@router.get("/catalog/albums/{album_id}")
async def catalog_album(album_id: str, storefront: str = "us"):
    """Fetch an album by catalog ID."""
    return await _apple_music_get(f"v1/catalog/{storefront}/albums/{album_id}")


@router.get("/catalog/playlists/{playlist_id}")
async def catalog_playlist(playlist_id: str, storefront: str = "us"):
    """Fetch a playlist by catalog ID."""
    return await _apple_music_get(f"v1/catalog/{storefront}/playlists/{playlist_id}")


@router.get("/me/storefront")
async def user_storefront(music_user_token: Optional[str] = Header(default=None, alias="Music-User-Token")):
    """Get the user's storefront based on their Music-User-Token."""
    if not music_user_token:
        raise HTTPException(status_code=400, detail="Music-User-Token header is required")
    return await _apple_music_get("v1/me/storefront", user_token=music_user_token)

File: apps/agent/src/apple-music.ts
/**
 * Apple Music API client with ES256 JWT developer token generation.
 * Ported from apps/backend/apple_music.py
 */
import * as jose from "jose";
import type { Env, TrackInfo } from "./types";

const APPLE_MUSIC_API_BASE = "https://api.music.apple.com";
const APPLE_MUSIC_TOKEN_MAX_TTL = 60 * 60 * 24 * 180; // 6 months

// Module-level token cache (Workers are single-threaded, safe)
let tokenCache: { token: string; exp: number } | null = null;

// ---------------------------------------------------------------------------
// Developer Token Generation
// ---------------------------------------------------------------------------

async function generateDeveloperToken(
  env: Env
): Promise<{ token: string; exp: number }> {
  const privateKey = await jose.importPKCS8(
    env.APPLE_MUSIC_PRIVATE_KEY.replace(/\\n/g, "\n"),
    "ES256"
  );
  const now = Math.floor(Date.now() / 1000);
  const ttl = Math.min(
    Math.max(parseInt(env.APPLE_MUSIC_TOKEN_TTL_SECONDS || "3600"), 60),
    APPLE_MUSIC_TOKEN_MAX_TTL
  );
  const exp = now + ttl;

  const token = await new jose.SignJWT({ iss: env.APPLE_MUSIC_TEAM_ID })
    .setProtectedHeader({ alg: "ES256", kid: env.APPLE_MUSIC_KEY_ID })
    .setIssuedAt(now)
    .setExpirationTime(exp)
    .sign(privateKey);

  return { token, exp };
}

export async function getDeveloperToken(
  env: Env
): Promise<{ token: string; exp: number }> {
  const now = Math.floor(Date.now() / 1000);

  if (tokenCache && now < tokenCache.exp - 60) {
    return tokenCache;
  }

  const result = await generateDeveloperToken(env);
  tokenCache = result;
  return result;
}

// ---------------------------------------------------------------------------
// Apple Music API Client
// ---------------------------------------------------------------------------

function buildHeaders(
  devToken: string,
  userToken?: string
): Record<string, string> {
  const headers: Record<string, string> = {
    Authorization: `Bearer ${devToken}`,
  };
  if (userToken) {
    headers["Music-User-Token"] = userToken;
  }
  return headers;
}

export async function appleMusicGet(
  path: string,
  env: Env,
  params?: Record<string, string | number>,
  userToken?: string
): Promise<Record<string, unknown>> {
  const { token } = await getDeveloperToken(env);
  const url = new URL(`/${path.replace(/^\//, "")}`, APPLE_MUSIC_API_BASE);

  if (params) {
    for (const [key, value] of Object.entries(params)) {
      url.searchParams.set(key, String(value));
    }
  }

  const response = await fetch(url.toString(), {
    headers: buildHeaders(token, userToken),
  });

  if (!response.ok) {
    const text = await response.text();
    throw new AppleMusicError(response.status, text);
  }

  return (await response.json()) as Record<string, unknown>;
}

class AppleMusicError extends Error {
  constructor(
    public status: number,
    public detail: string
  ) {
    super(`Apple Music API error ${status}: ${detail}`);
  }
}

// ---------------------------------------------------------------------------
// Parse track from Apple Music API response
// ---------------------------------------------------------------------------

export function parseTrackFromSong(song: Record<string, unknown>): TrackInfo {
  const attrs = (song.attributes || {}) as Record<string, unknown>;
  const artwork = (attrs.artwork || {}) as Record<string, unknown>;
  return {
    id: song.id as string,
    name: (attrs.name as string) || "Unknown",
    artist: (attrs.artistName as string) || "Unknown Artist",
    album: attrs.albumName as string | undefined,
    artwork_url: artwork.url as string | undefined,
    duration: attrs.durationInMillis
      ? (attrs.durationInMillis as number) / 1000
      : undefined,
  };
}

// ---------------------------------------------------------------------------
// Route Handler for Apple Music endpoints
// ---------------------------------------------------------------------------

export async function handleAppleMusic(
  request: Request,
  env: Env
): Promise<Response> {
  const url = new URL(request.url);
  const path = url.pathname.replace(/^\/apple-music/, "");

  try {
    // GET /apple-music/developer-token
    if (path === "/developer-token") {
      const { token, exp } = await getDeveloperToken(env);
      return Response.json({ token, expires_at: exp });
    }

    // GET /apple-music/catalog/search?term=...&types=...&storefront=...&limit=...&offset=...
    if (path === "/catalog/search") {
      const term = url.searchParams.get("term");
      if (!term) {
        return Response.json({ error: "term required" }, { status: 400 });
      }
      const storefront = url.searchParams.get("storefront") || "us";
      const result = await appleMusicGet(
        `v1/catalog/${storefront}/search`,
        env,
        {
          term,
          types: url.searchParams.get("types") || "songs",
          limit: parseInt(url.searchParams.get("limit") || "10"),
          offset: parseInt(url.searchParams.get("offset") || "0"),
        }
      );
      return Response.json(result);
    }

    // GET /apple-music/catalog/songs/:id
    const songsMatch = path.match(/^\/catalog\/songs\/(.+)$/);
    if (songsMatch) {
      const storefront = url.searchParams.get("storefront") || "us";
      const result = await appleMusicGet(
        `v1/catalog/${storefront}/songs/${songsMatch[1]}`,
        env
      );
      return Response.json(result);
    }

    // GET /apple-music/catalog/albums/:id
    const albumsMatch = path.match(/^\/catalog\/albums\/(.+)$/);
    if (albumsMatch) {
      const storefront = url.searchParams.get("storefront") || "us";
      const result = await appleMusicGet(
        `v1/catalog/${storefront}/albums/${albumsMatch[1]}`,
        env
      );
      return Response.json(result);
    }

    // GET /apple-music/catalog/playlists/:id
    const playlistsMatch = path.match(/^\/catalog\/playlists\/(.+)$/);
    if (playlistsMatch) {
      const storefront = url.searchParams.get("storefront") || "us";
      const result = await appleMusicGet(
        `v1/catalog/${storefront}/playlists/${playlistsMatch[1]}`,
        env
      );
      return Response.json(result);
    }

    // GET /apple-music/me/storefront
    if (path === "/me/storefront") {
      const userToken = request.headers.get("Music-User-Token");
      if (!userToken) {
        return Response.json(
          { error: "Music-User-Token header is required" },
          { status: 400 }
        );
      }
      const result = await appleMusicGet(
        "v1/me/storefront",
        env,
        undefined,
        userToken
      );
      return Response.json(result);
    }

    // GET /apple-music/validate-token?user_id=...
    if (path === "/validate-token") {
      const userId = url.searchParams.get("user_id");
      if (!userId) {
        return Response.json({ error: "user_id required" }, { status: 400 });
      }

      const { results } = await env.DB.prepare(
        'SELECT "appleMusicToken" FROM "profile" WHERE "id" = ?'
      )
        .bind(userId)
        .all();

      if (
        !results.length ||
        !(results[0] as Record<string, unknown>).appleMusicToken
      ) {
        return Response.json({ valid: false, reason: "no_token" });
      }

      const userToken = (results[0] as Record<string, unknown>)
        .appleMusicToken as string;
      try {
        await appleMusicGet("v1/me/storefront", env, undefined, userToken);
        return Response.json({ valid: true });
      } catch (e) {
        if (e instanceof AppleMusicError && (e.status === 401 || e.status === 403)) {
          await env.DB.prepare(
            'UPDATE "profile" SET "appleMusicToken" = NULL WHERE "id" = ?'
          )
            .bind(userId)
            .run();
          return Response.json({ valid: false, reason: "token_expired" });
        }
        throw e;
      }
    }

    return Response.json({ error: "Not found" }, { status: 404 });
  } catch (e) {
    if (e instanceof AppleMusicError) {
      return Response.json(
        { error: e.detail, status: e.status },
        { status: e.status }
      );
    }
    console.error("Apple Music error:", e);
    return Response.json({ error: String(e) }, { status: 500 });
  }
}
